text/3424-cargo-script.md

```diff
@@ -0,0 +1,585 @@
+- Feature Name: cargo-script
+- Start Date: 2023-04-26
+- Pre-RFC: [internals](https://internals.rust-lang.org/t/pre-rfc-cargo-script-for-everyone/18639)
+- eRFC PR: [rust-lang/rfcs#3424](https://github.com/rust-lang/rfcs/pull/3424)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This *experimental RFC* adds unstable support for single-file
+packages in cargo so we can explore the design and resolve
+questions with an implementation to collect feedback on.
+
+Single-file packages are `.rs` files with an embedded
+manifest.  These will be accepted with just like `Cargo.toml` files with
+`--manifest-path`.  `cargo` will be modified to accept `cargo <file>.rs` as a
+shortcut to `cargo run --manifest-path <file>.rs`.  This allows placing
+`cargo` in a `#!` line for directly running these files.
+
+Example:
+```rust
+#!/usr/bin/env cargo
+
+//! ```cargo
+//! [dependencies]
+//! clap = { version = "4.2", features = ["derive"] }
+//! ```
+
+use clap::Parser;
+
+#[derive(Parser, Debug)]
+#[clap(version)]
+struct Args {
+    #[clap(short, long, help = "Path to config")]
+    config: Option<std::path::PathBuf>,
+}
+
+fn main() {
+    let args = Args::parse();
+    println!("{:?}", args);
+}
+```
+```console
+$ ./prog --config file.toml
```

Comment:
Does this console invocation need to have the `.rs` extension on the filename?
```suggestion
$ ./prog.rs --config file.toml
```
I was following along earlier in the pre-rfc thread on IRLO, but wasn't sure whether the `.rs` extension was still an unresolved question. (I see the other `hello_world.rs` example does have the `.rs` filename extension)

Reply:
> Does this console invocation need to have the .rs extension on the filename?

Either one is fine here.  For the utility workflows, people don't always like to put extensions on their scripts (independent of the language). 

Granted, that can affect what we do for `cargo test --manifest-path ./prog`.  We might not be able to key just off of the extension.

> I was following along earlier in the pre-rfc thread on IRLO, but wasn't sure whether the .rs extension was still an unresolved question. (I see the other hello_world.rs example does have the .rs filename extension)

Most of that extension discussion is about what extension would be used when it is present and not so much on the use cases for when to drop it.

Reply:
> Either one is fine here.

How is this supposed to work in general? If I name the file `run`, then cargo will be invoked as `cargo run`, which looks like the subcommand and not like a request to execute a file. `cargo <file>.rs` is at least unambiguous.

Reply:
The exact precedence rules are left to be explored with the implementation.  This is recorded in "Unresolved questions" as

> If cargo, what precedence does it have compared to built-in commands, aliases, and external commands.

In epage/cargo-script-mvs#154, I've started looking into prior art among other build tools that support `<tool> <file>`.

dlang's `dub` is of particular interest
1. If the `<positional>` is `-` or ends with `.rs`, treat it as a single-file package
2. If `<positional>` matches an existing subcommand, run it
3. If `<positional>` exists, treat it as a single-file package
4. If `<positional>.d` exists, treat it as a single-file package
5. Otherwise, treat `<positional>` as a non-existent subcommand

For me, some tweaks I can see making
- Don't support "`<positional>.d` exists" check (4)
- For (1), check if `<positional>` has path separators (at least on my system, `#!` causes an absolute path to be passed to the command)

Reply:
> * For (1), check if `<positional>` has path separators (at least on my system, `#!` causes an absolute path to be passed to the command)

it doesn't on my system (Android).

Reply:
Anything involving paths is likely to be system specific. Or at least they'll be edge case systems.

Reply:
> Either one is fine here. For the utility workflows, people don't always like to put extensions on their scripts (independent of the language).

People can just use a symlink though. In the case of the PHP PHAR format, PHP denies loading phar files that do not have a `.phar` extension for some reason (https://github.com/php/php-src/blob/a65cdd97a148dccdb68a5f58c258d54c305f1b2e/ext/phar/phar.c#L1961), so some installations of composer are performed by `ln -s composer.phar /usr/bin/composer` (citation needed: I know several people who do so, but I am not sure if this is a common practice).

Reply:
> If I name the file `run`, then cargo will be invoked as `cargo run`, which looks like the subcommand and not like a request to execute a file.

`cargo -- run`?

Does this work?

 `#!/usr/bin/env cargo -- --`